// thegamesdb/lib/api.js
'use strict';

var axios = require('axios');
var parse = require('./utils').parseXml;
var metadata = require('./metadata');
var reducers = require('./reducers')(metadata);

module.exports = {
  getGamesList: function getGamesList(args) {
    // Raise if name is not declared
    if (!args.name) throw "'name' argument is required";

    return axios.get(metadata.endpoint + '/GetGamesList.php', { params: args }).then(function (response) {
      return parse(response.data);
    }).then(function (data) {
      return data.Data.Game || [];
    }).then(reducers.gameList);
  },
  getGame: function getGame(args) {
    // Raise if id is not declared
    if (!args.id) throw "'id' argument is required";

    return axios.get(metadata.endpoint + '/GetGame.php', { params: args }).then(function (response) {
      return parse(response.data);
    }).then(function (data) {
      return data.Data.Game[0] || {};
    }).then(reducers.gameDetail);
  },
  getGameArt: function getGameArt(args) {
    // Raise if id is not declared
    if (!args.id) throw "'id' argument is required";

    return axios.get(metadata.endpoint + '/GetArt.php', { params: args }).then(function (response) {
      return parse(response.data);
    }).then(function (data) {
      return data.Data.Images || [];
    }).then(reducers.gameArtList);
  },
  getPlatformsList: function getPlatformsList() {
    return axios.get(metadata.endpoint + '/GetPlatformsList.php').then(function (response) {
      return parse(response.data);
    }).then(function (data) {
      return data.Data ? data.Data.Platforms[0].Platform : [];
    }).then(reducers.platformList);
  },
  getPlatform: function getPlatform(args) {
    // Raise if id is not declared
    if (!args.id) throw "'id' argument is required";

    return axios.get(metadata.endpoint + '/GetPlatform.php', { params: args }).then(function (response) {
      return parse(response.data);
    }).then(function (data) {
      return data.Data.Platform[0] || {};
    }).then(reducers.platformDetail);
  },
  getPlatformGames: function getPlatformGames(args) {
    // Raise if id is not declared
    if (!args.id) throw "'id' argument is required";
    var params = { platform: args.id };

    return axios.get(metadata.endpoint + '/GetPlatformGames.php', { params: params }).then(function (response) {
      return parse(response.data);
    }).then(function (data) {
      return data.Data.Game || [];
    }).then(reducers.gameList);
  }
};
